sbi/inference/snpe/snpe_base.py

```diff
@@ -155,81 +145,80 @@ def __call__(
-            # things to tensorboard anymore. Calling `leakage_correction()` is needed
-            # to update the leakage after each round.
-            if self._posterior.default_x is None:
-                acceptance_rate = torch.tensor(float("nan"))
-            else:
-                acceptance_rate = self._posterior.leakage_correction(
-                    x=self._posterior.default_x,
-                    force_update=True,
-                    show_progress_bars=self._show_progress_bars,
-                )
+        # Fit posterior using newly aggregated data set.
+        self._train(
+            training_batch_size=training_batch_size,
+            learning_rate=learning_rate,
+            validation_fraction=validation_fraction,
+            stop_after_epochs=stop_after_epochs,
+            max_num_epochs=cast(int, max_num_epochs),
+            clip_max_norm=clip_max_norm,
+            calibration_kernel=calibration_kernel,
+            exclude_invalid_x=exclude_invalid_x,
+            discard_prior_samples=discard_prior_samples,
+        )
 
-            # Update tensorboard and summary dict.
-            self._summarize(
-                round_=round_,
-                x_o=self._posterior.default_x,
-                theta_bank=theta,
-                x_bank=x,
-                posterior_samples_acceptance_rate=acceptance_rate,
+        # Store models at end of each round.
+        self._model_bank.append(deepcopy(self._posterior))
+        self._model_bank[-1].net.eval()
+
+        # Making the call to `leakage_correction()` and the update of
+        # self._leakage_density_correction_factor explicit here. This is just
+        # to make sure this update never gets lost when we e.g. do not log our
+        # things to tensorboard anymore. Calling `leakage_correction()` is needed
+        # to update the leakage after each round.
+        if self._posterior.default_x is None:
+            acceptance_rate = torch.tensor(float("nan"))
+        else:
+            acceptance_rate = self._posterior.leakage_correction(
+                x=self._posterior.default_x,
+                force_update=True,
+                show_progress_bars=self._show_progress_bars,
             )
 
-            # Update description for progress bar.
-            if self._show_round_summary:
-                print(self._describe_round(round_, self._summary))
+        # Update tensorboard and summary dict.
+        self._summarize(
+            round_=self._round,
+            x_o=self._posterior.default_x,
+            theta_bank=theta,
+            x_bank=x,
+            posterior_samples_acceptance_rate=acceptance_rate,
+        )
+
+        # Update description for progress bar.
+        if self._show_round_summary:
+            print(self._describe_round(self._round, self._summary))
 
-        self._posterior._num_trained_rounds = num_rounds
+        self._posterior._num_trained_rounds = self._round + 1
```

Comment:
When is this property used?

Reply:
Only for a warning in SRE. It needs to know the number of trained rounds when evaluating log_prob(): if single-round, no warning. If multi-round, warn that it's only correct up to a normalizing constant